controller/resource.go

```diff
@@ -343,6 +352,66 @@ func (s TaskStateLister) createNamespaces(resources []Resource) error {
 	return result
 }
 
+// ReconcileServices ensures services are deregistered for missing ECS tasks with the
+// following steps
+//
+//   - Query the list of services belonging to the clusterARN node from Consul
+//   - Filter out the services that are registered by a Consul ECS control plane container
+//     running in a mesh task without a Consul client container.
+//   - Deregister those services for which the backing task resource is not available in ECS.
+func (s TaskStateLister) ReconcileServices(resources []Resource) error {
```

Comment:
Can we have `TaskStateLister.List()` discover service instances instead of adding `ReconcileServices` to the interface?

That should fit into the existing interfaces:

* `TaskStatelist.List()` produces a list of `Resource`s (which are `TaskState`s) to reconcile.
   * A TaskState represents an ECS task and should have all the fields necessary to know the ACL tokens associated with the task, the service ids in Consul associated with the task, and whether those need to be cleaned up (i.e. whether the task is still running)
* Then each `Resource` can be independently reconciled - to delete tokens and service instances

I think we have `ReconcileNamespaces` as an exception because there is a `many:few` mapping of Resources to Namespaces, so it would be inefficient to reconcile namespaces as part of each TaskState.Reconcile.

But in this case, each task directly maps to a couple of service instances and those can be independently reconciled, and I don't think we need to deviate from the existing interfaces.

What do you think?

Reply:
Yeah. This makes sense and looks a lot more cleaner in terms of responsibilities. Thanks for the feedback